### PR TITLE
[MAD-16896] UITextInputComponent improvements

### PR DESCRIPTION
**Pull request by carbfeng**
Ticket 16896

## What does this PR do?

Support showing the native iOS text input field on top of the game text field.
Turn on autocorrect for the iOS text input field.
Hides the game text field when text entry starts.
Support changing the return key to show Send on the virtual keyboard.

## How was this PR tested?

Tested on an iOS device build

### CODE REVIEW

File: Code/Framework/AzFramework/AzFramework/Input/Buses/Requests/InputTextEntryRequestBus.h [hunk 1]
```diff
@@ -30,6 +30,13 @@ namespace AzFramework
         {
             AZStd::string m_initialText;   //!< The virtual keyboard's initial text
             AZStd::string m_titleText;     //!< The virtual keyboard's title text
+#if defined(CARBONATED)
+            bool m_showSendOnReturnKey = false; //!< Whether to show the send button on the return key
```
Comment (carbonated-niko) [inline Code/Framework/AzFramework/AzFramework/Input/Buses/Requests/InputTextEntryRequestBus.h]: having a comment that matches the bool name doesnt really help dev understanding - If you could lets change the comment to be something like // Whether we set the native keyboard return key type to Send

File: Gems/LyShine/Code/Source/UiTextInputComponent.h [hunk 2]
```diff
@@ -211,4 +211,8 @@ class UiTextInputComponent
     bool m_clipInputText;               //!< True if input text should be visually clipped to child text element, false otherwise.
 
     bool m_enableClipboard;             //!< True if copy/cut/paste should be supported, false otherwise.
+
+#if defined(CARBONATED)
+    bool m_showSendOnReturnKey;         //!< True if return key on virtual keyboard should show send, false otherwise.
```
Comment (carbonated-niko) [inline Gems/LyShine/Code/Source/UiTextInputComponent.h:220]: oh like this comment

**Review by carbonated-niko — APPROVED** [review]

File: Code/Framework/AzFramework/AzFramework/Input/Buses/Requests/InputTextEntryRequestBus.h (hunk 1 above)
Comment (carbfeng) [reply Code/Framework/AzFramework/AzFramework/Input/Buses/Requests/InputTextEntryRequestBus.h]: Changed comment

File: Gems/LyShine/Code/Source/UiTextInputComponent.cpp [hunk 3]
```diff
@@ -108,6 +108,7 @@ namespace
         utf8String.erase(left, right - left);
     }
 
+#if !defined(CARBONATED) || !defined(AZ_PLATFORM_IOS)
```
Comment (JasonCarbonated) [inline Gems/LyShine/Code/Source/UiTextInputComponent.cpp]: This would be more idiomatic if expressed without the use of not operators when possible. It can be confusing when using !a || !b
Consider using:
`#if defined(CARBONATED) && defined(AZ_PLATFORM_IOS)`
instead

File: Gems/LyShine/Code/Source/UiTextInputComponent.cpp [hunk 4]
```diff
@@ -332,6 +337,14 @@ bool UiTextInputComponent::HandleTextInput(const AZStd::string& inputTextUTF8)
 
     bool changedText = false;
 
+#if defined(CARBONATED) && defined(AZ_PLATFORM_IOS)
```
Comment (JasonCarbonated) [inline Gems/LyShine/Code/Source/UiTextInputComponent.cpp]: yeah! Like this one -- this one is good.

**Review by JasonCarbonated — CHANGES_REQUESTED** [review]

File: Gems/LyShine/Code/Source/UiTextInputComponent.cpp (hunk 3 above)
Comment (carbfeng) [reply Gems/LyShine/Code/Source/UiTextInputComponent.cpp]: Changed to 
`!(defined(CARBONATED) && defined(AZ_PLATFORM_IOS))`

**Review by JasonCarbonated — APPROVED** [review]

**Review by Twolewis — APPROVED** [review]

File: Gems/LyShine/Code/Source/UiTextInputComponent.cpp [hunk 5]
```diff
@@ -108,6 +108,7 @@ namespace
         utf8String.erase(left, right - left);
     }
 
+#if !(defined(CARBONATED) && defined(AZ_PLATFORM_IOS))
```
Comment (Twolewis) [inline Gems/LyShine/Code/Source/UiTextInputComponent.cpp]: Why was this removed?

File: Gems/LyShine/Code/Source/UiTextInputComponent.cpp (hunk 4 above)
Comment (Twolewis) [inline Gems/LyShine/Code/Source/UiTextInputComponent.cpp]: O3DE (rightfully) doesn't like platform-specific code in platform-agnostic engine code.  The UiTextInputComponent being a good example - it should not care about ios or android or whatever.  You'll need to remove the AZ_PLATFORM_IOS guard here and make this code platform agnostic.

**Review by Twolewis — CHANGES_REQUESTED** [review]

File: Gems/LyShine/Code/Source/UiTextInputComponent.cpp (hunk 5 above)
Comment (carbfeng) [reply Gems/LyShine/Code/Source/UiTextInputComponent.cpp]: Because it was causing an unused function error. I've removed the preprocessor defines

Comment (Twolewis) [reply Gems/LyShine/Code/Source/UiTextInputComponent.cpp]: That's odd.  We have many unused functions that don't throw errors.  Remove this guard to invoke the error and let's look at it together.

File: Gems/LyShine/Code/Source/UiTextInputComponent.cpp (hunk 4 above)
Comment (carbfeng) [reply Gems/LyShine/Code/Source/UiTextInputComponent.cpp]: Removed all the AZ_PLATFORM defines in UiTextInputComponent

File: Gems/LyShine/Code/Source/UiTextInputComponent.cpp (hunk 5 above)
Comment (carbfeng) [reply Gems/LyShine/Code/Source/UiTextInputComponent.cpp]: I no longer get the error after removing all the AZ_PLATFORM_IOS defines

**Review by Twolewis — APPROVED** [review]
